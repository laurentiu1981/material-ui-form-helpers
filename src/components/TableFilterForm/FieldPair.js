import React from "react";
import {Field} from "redux-form";
import { makeStyles } from "@material-ui/styles";
import ClearIcon from '@material-ui/icons/Clear';
import {renderField} from "../../renderField";

const useStyle = makeStyles({
  container: {
    margin: '5px',
    display: 'flex',
    position: 'relative',
    paddingTop: '30px'
  },
  field: {
    margin: '5px'
  }
});


export const FieldPair = ({field, defaultValue, options, removeCallback, index}) => {
  const classes = useStyle();
  return (
    <div className={`${index}-sort-pair`}>
        <div>
          <Field
            name={`${field}.fieldName`}
            type="chosenSelect"
            component={renderField}
            options={options}
            label="Sort"
            className={classes.field}
          />
          <Field
            name={`${field}.direction`}
            type="chosenSelect"
            component={renderField}
            options={[{label: "ASC", key: "ASC"},{label: "DESC", key: "DESC"} ]}
            label="Direction"
            className={classes.field}
          />
          <ClearIcon className={`${index}-sort-pair-clear-icon`} onClick={removeCallback} />
        </div>
    </div>
  )
};