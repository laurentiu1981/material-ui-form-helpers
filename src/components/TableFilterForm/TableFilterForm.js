import React from 'react';
import Chip from "@material-ui/core/Chip";
import {renderField} from "../../renderField";
import {Field, reduxForm, change, propTypes, arrayRemove } from "redux-form";
import Button from "@material-ui/core/Button";
import Paper from "@material-ui/core/Paper";
import Grid from "@material-ui/core/Grid";
import { withStyles } from '@material-ui/core/styles';
import SubmitButton from "../SubmitButton";
import DehazeIcon from '@material-ui/icons/Dehaze';
import FormWrapper from "../FormWrapper";
import {connect} from "react-redux";
import PropTypes from 'prop-types';
import _ from 'lodash';
import has from 'lodash/has';

const styles = () => ({
  chipsContainer: {
    marginTop: "0.5em",
    padding: "0.25em 0.5em",
  },
  actionsContainer: {
    marginTop: "0.5em",
  }
});

const FilterFormWrapper = withStyles(theme => ({
  root: {
    width: '100%',
    marginTop: 0,
    position: "relative",
    borderBottomLeftRadius: 0,
    borderBottomRightRadius: 0,
  },
  formRoot: {
    padding: "1em"
  },
}))(FormWrapper);

const FilterFormField = (props) => {
  const {xs, sm, md, lg, name, label, type } = props;
  switch (type) {
    case 'min-max':
      return (
        <React.Fragment>
          <Grid item xs={xs} sm={sm} md={md} lg={lg}>
            <Field
              type="text"
              component={renderField}
              name={`${name}_min`}
              key={`${label}_min`}
              label={`${label} Min`}
            >
              {`${label} Min`}
            </Field>
          </Grid>
          <Grid item xs={xs} sm={sm} md={md} lg={lg}>
            <Field
              type="text"
              component={renderField}
              name={`${name}_max`}
              key={`${label}_max`}
              label={`${label} Max`}
            >
              {`${label} Max`}
            </Field>
          </Grid>
        </React.Fragment>
      );
    default:
      return (
        <Grid item xs={xs} sm={sm} md={md} lg={lg}>
          <Field
            type="text"
            component={renderField}
            name={name}
            key={label}
            label={label}
          >
            {label}
          </Field>
        </Grid>
      )
  }
};

FilterFormField.propTypes = {
  name: PropTypes.string,
  label: PropTypes.string,
  type: PropTypes.string,
  xs: PropTypes.number,
  sm: PropTypes.number,
  md: PropTypes.number,
  lg: PropTypes.number,
};

FilterFormField.defaultProps = {
  xs: 12,
  sm: 6,
  md: 3,
  lg: 2,
}

class TableFilterForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      isAdvanced: false,
      submittedData: {},
    };
  }

  renderField = field => (
    <FilterFormField
      name={field.name}
      key={field.label}
      label={field.label}
      type={field.filter.type}
      {...(field.filter.size ? field.filter.size : {})}
    >
      {field.label}
    </FilterFormField>
  );

  renderFilters = () => {
    return this.props.tableDefinition
      .filter((field) => {
        return field.hasOwnProperty("filter") &&
          field.filter.hasOwnProperty("type") &&
          (field.filter.hasOwnProperty("basicFilter") || this.state.isAdvanced)
      })
      .map(this.renderField);
  }

  hasAdvancedFilters = () => {
    return !!this.props.tableDefinition
      .filter((field) => {
        return has(field, ['filter', 'type']) &&
          !has(field, ['filter', 'basicFilter'])
      }).length
  }

  renderChips = (submittedData) => {
    const {classes} = this.props;
    let filters = Object.keys(submittedData);
    if (filters.length) {
      return (
        <Paper className={classes.chipsContainer}>
          {filters.map((key) => {
            return (
              <Chip
                key={key}
                label={`${key}: ${_.isObject(submittedData[key]) ? Object.values(submittedData[key]) : submittedData[key]}`}
                onDelete={this.handleChipDeleteForFilter(key)}
                color="primary"
              />
            )
          })}
        </Paper>
      )
    }
    return null;
  };

  onClickExpandHandler = (event) => {
    event.preventDefault();
    this.setState({
      isAdvanced: !this.state.isAdvanced
    })
  };

  handleChipDeleteForFilter = (keyChipToDelete) => () => {
    if (this.state.submittedData.hasOwnProperty(keyChipToDelete)) {
      const submittedData = _.omit(this.state.submittedData, keyChipToDelete);
      this.props.changeFieldValue(keyChipToDelete, '');
      this.setState({ submittedData: submittedData },
        () => {
          this.props.onSubmit(this.state.submittedData);
        });
    }
  };

  onSubmit = async (formValues) => {
    await this.props.onSubmit(formValues);
    this.setState({submittedData: formValues});
  };

  onCancel = () => {
    this.props.destroy('TableFilterForm');
    this.setState({submittedData: {}});
    if (this.props.onResetCallback) {
      this.props.onResetCallback();
    }
  };

  render() {
    const {submitting, classes} = this.props;
    const isSubmitting = submitting ? 1 : 0;
    return (
      <FilterFormWrapper xs={12} sm={12} md={12} lg={12}>
        <form onSubmit={this.props.handleSubmit(this.onSubmit)} autoComplete="off">
          <Grid
            container
            spacing={1}
            direction="row"
            alignItems="flex-start"
            justify="center"
          >
            <Grid container spacing={1}>
              {this.renderFilters()}
            </Grid>

            {this.hasAdvancedFilters() &&
              <Grid container spacing={1} justify="flex-end">
                <Grid item>
                  <Button onClick={this.onClickExpandHandler}><DehazeIcon />{this.state.isAdvanced ? "Hide Advanced Filters" : "Show Advanced Filters"}</Button>
                </Grid>
              </Grid>
            }
          </Grid>

          {this.renderChips(this.state.submittedData)}

          <div className={classes.actionsContainer}>
            <SubmitButton type="submit" variant="contained" color="primary" is_submitting={isSubmitting}
                          disabled={!!submitting}>
              Filter
            </SubmitButton>
            <Button onClick={this.onCancel}>
              Cancel
            </Button>
          </div>
        </form>
      </FilterFormWrapper>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    changeFieldValue: (field, value) => {
      dispatch(change('TableFilterForm', field, value))
    },
    arrayRemoveElement: (field, index) => {
      dispatch(arrayRemove('TableFilterForm', field, index))
    }
  }
};

TableFilterForm.propTypes = {
  ...propTypes,
  onResetCallback: PropTypes.func,
  handleSubmit: PropTypes.func,
  onSubmit: PropTypes.func,
  tableDefinition: PropTypes.array,
};

const TableFilterFormRedux = reduxForm({
  form: 'TableFilterForm'
})(TableFilterForm);

export default withStyles(styles)(connect(null, mapDispatchToProps)(TableFilterFormRedux));
