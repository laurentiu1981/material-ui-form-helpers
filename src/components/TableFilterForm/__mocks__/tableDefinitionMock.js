export const tableDefinition = [
  {
    label: "Test simple element",
    property: "simpleElement",
    sort: {
      defaultSort: "ASC",
    },
    filter: {
      name: 'simpleElement',
      type: 'text',
      dataType: 'text',
      weight: 2,
      basicFilter: true,
    },
    name: 'simpleElement',
  },
  {
    label: "Test interval element",
    property: "intervalElement",
    sort: true,
    filter: {
      name: 'intervalElement',
      type: 'min-max',
      dataType: 'date',
      weight: 2,
    },
    name: 'intervalElement',
  },
];