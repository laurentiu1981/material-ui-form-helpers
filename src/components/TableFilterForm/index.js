import TableFilterForm from './TableFilterForm';
export default TableFilterForm;