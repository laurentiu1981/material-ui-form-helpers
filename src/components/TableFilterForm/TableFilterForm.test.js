import React from 'react';
import { combineReducers, createStore } from 'redux';
import {fireEvent, render} from '@testing-library/react'
import {reducer as form} from 'redux-form';
import TableFilterForm from './TableFilterForm';
import {Provider} from "react-redux";
import '@testing-library/jest-dom/extend-expect';
import {createMuiTheme, MuiThemeProvider} from "@material-ui/core";
import {tableDefinition} from "./__mocks__/tableDefinitionMock";

const theme = createMuiTheme({});

const defaultSortsMock = (tableDefinition) => {
  return tableDefinition.filter((field) => field.hasOwnProperty('sort') && field.sort.hasOwnProperty('defaultSort')).map((field) => {
    return {
      fieldName: {value: field.name, label: field.label},
      direction: {value: field.sort.defaultSort, label: field.sort.defaultSort},
    };
  });
};

describe('TableFilterForm Tests', () => {

  let store;

  beforeEach(() => {
    store = createStore(combineReducers({form}));
  });

  test('Table Filter Form - submit', (done) => {

    const onSubmitPropMock = jest.fn().mockImplementation(() => {
      return {};
    });

    const utils = render(
      <Provider store={store}>
        <MuiThemeProvider theme={theme}>
          <TableFilterForm onSubmit={onSubmitPropMock} tableDefinition={tableDefinition}
                           defaultSort={defaultSortsMock(tableDefinition)}/>
        </MuiThemeProvider>
      </Provider>
    );

    // default filter field
    const defaultFilterField = utils.getByText('Test simple element');
    expect(defaultFilterField).toBeInTheDocument();

    // chip for default filter
    let defaultSortChip = utils.getByText('Test simple element: ASC');
    expect(defaultSortChip).toBeInTheDocument();

    const simpleElementInput = utils.container.querySelector('input[name="simpleElement"]');
    fireEvent.change(simpleElementInput, {target: {value: 'testInputSimpleElement'}});
    expect(simpleElementInput).toMatchSnapshot();
    expect(simpleElementInput.value).toBe('testInputSimpleElement');

    // display advanced filters
    fireEvent.click(utils.getByText('Advanced Filters'));

    const intervalElementMin = utils.container.querySelector('input[name="intervalElement_min"]');
    fireEvent.change(intervalElementMin, {target: {value: "1"}});
    expect(intervalElementMin).toMatchSnapshot();
    expect(intervalElementMin.value).toBe("1");

    const intervalElementMax = utils.container.querySelector('input[name="intervalElement_max"]');
    fireEvent.change(intervalElementMax, {target: {value: "10"}});
    expect(intervalElementMax).toMatchSnapshot();
    expect(intervalElementMax.value).toBe("10");

    const formComponent = utils.container.querySelector('form');

    fireEvent.submit(formComponent);
    expect(onSubmitPropMock).toHaveBeenCalled();

    // hide advanced filters
    fireEvent.click(utils.getByText('Advanced Filters'));

    setTimeout(() => {
      // check chips values
      const chipSimpleElementFilter = utils.queryByText("simpleElement: testInputSimpleElement");
      expect(chipSimpleElementFilter).toBeInTheDocument();
      const chipIntervalElementMin = utils.queryByText("intervalElement_min: 1");
      expect(chipIntervalElementMin).toBeInTheDocument();
      const chipIntervalElementMax = utils.queryByText("intervalElement_max: 10");
      expect(chipIntervalElementMax).toBeInTheDocument();
      const defaultSortChip = utils.queryByText('Test simple element: ASC');
      expect(defaultSortChip).toBeInTheDocument();
      done();
    }, 1000);
  });

  test('Table Filter Form - submit & cancel', (done) => {

    const onSubmitPropMock = jest.fn().mockImplementation(() => {
      return {};
    });

    const onCancelPropMock = jest.fn().mockImplementation(() => {
      return {};
    });

    const utils = render(
      <Provider store={store}>
        <MuiThemeProvider theme={theme}>
          <TableFilterForm onSubmit={onSubmitPropMock} tableDefinition={tableDefinition}
                           defaultSort={defaultSortsMock(tableDefinition)} onCancel={onCancelPropMock}/>
        </MuiThemeProvider>
      </Provider>
    );

    const simpleElementInput = utils.container.querySelector('input[name="simpleElement"]');
    fireEvent.change(simpleElementInput, {target: {value: 'testInputSimpleElement'}});
    expect(simpleElementInput).toMatchSnapshot();
    expect(simpleElementInput.value).toBe('testInputSimpleElement');

    // display advanced filters
    fireEvent.click(utils.getByText('Advanced Filters'));

    const intervalElementMin = utils.container.querySelector('input[name="intervalElement_min"]');
    fireEvent.change(intervalElementMin, {target: {value: "1"}});
    expect(intervalElementMin).toMatchSnapshot();
    expect(intervalElementMin.value).toBe("1");

    const intervalElementMax = utils.container.querySelector('input[name="intervalElement_max"]');
    fireEvent.change(intervalElementMax, {target: {value: "10"}});
    expect(intervalElementMax).toMatchSnapshot();
    expect(intervalElementMax.value).toBe("10");

    const formComponent = utils.container.querySelector('form');

    fireEvent.submit(formComponent);
    expect(onSubmitPropMock).toHaveBeenCalled();

    // hide advanced filters
    fireEvent.click(utils.getByText('Advanced Filters'));

    setTimeout(() => {

      // check chips values
      let chipSimpleElement = utils.queryByText("simpleElement: testInputSimpleElement");
      expect(chipSimpleElement).toBeInTheDocument();
      let chipIntervalElementMin = utils.queryByText("intervalElement_min: 1");
      expect(chipIntervalElementMin).toBeInTheDocument();
      let chipIntervalElementMax = utils.queryByText("intervalElement_max: 10");
      expect(chipIntervalElementMax).toBeInTheDocument();
      let defaultSortChip = utils.queryByText("Test simple element: ASC");
      expect(defaultSortChip).toBeInTheDocument();

      // clear fields & delete chips
      fireEvent.click(utils.getByText('Cancel'));

      // display advanced filters
      fireEvent.click(utils.getByText('Advanced Filters'));

      // check clear fields
      const simpleElementInput = utils.container.querySelector('input[name="simpleElement"]');
      expect(simpleElementInput).toMatchSnapshot();
      expect(simpleElementInput.value).toBe('');
      const intervalElementMin = utils.container.querySelector('input[name="intervalElement_min"]');
      expect(intervalElementMin).toMatchSnapshot();
      expect(intervalElementMin.value).toBe('');
      const intervalElementMax = utils.container.querySelector('input[name="intervalElement_max"]');
      expect(intervalElementMax).toMatchSnapshot();
      expect(intervalElementMax.value).toBe('');

      // hide advanced filters
      fireEvent.click(utils.getByText('Advanced Filters'));

      // check delete chips
      chipSimpleElement = utils.queryByText("simpleElement: testInputSimpleElement");
      expect(chipSimpleElement).toBeNull();
      chipIntervalElementMin = utils.queryByText("intervalElement_min: 1");
      expect(chipIntervalElementMin).toBeNull();
      chipIntervalElementMax = utils.queryByText("intervalElement_max: 10");
      expect(chipIntervalElementMax).toBeNull();
      defaultSortChip = utils.queryByText("Test simple element: ASC");
      expect(defaultSortChip).toBeNull();

      done();
    }, 1000);
  });

  test('Table Filter Form - add & remove sort pair', () => {

    const onSubmitPropMock = jest.fn().mockImplementation(() => {
      return {};
    });

    const utils = render(
      <Provider store={store}>
        <MuiThemeProvider theme={theme}>
          <TableFilterForm onSubmit={onSubmitPropMock} tableDefinition={tableDefinition}
                           defaultSort={defaultSortsMock(tableDefinition)}/>
        </MuiThemeProvider>
      </Provider>
    );

    // display advanced filters
    fireEvent.click(utils.getByText('Advanced Filters'));

    const intervalElementMin = utils.container.querySelector('input[name="intervalElement_min"]');
    fireEvent.change(intervalElementMin, {target: {value: "1"}});
    expect(intervalElementMin).toMatchSnapshot();
    expect(intervalElementMin.value).toBe("1");

    const intervalElementMax = utils.container.querySelector('input[name="intervalElement_max"]');
    fireEvent.change(intervalElementMax, {target: {value: "10"}});
    expect(intervalElementMax).toMatchSnapshot();
    expect(intervalElementMax.value).toBe("10");

    var elementsWithFormItemClass = utils.container.querySelectorAll(".form-item");
    var lengthElementsWithFormItemClass = elementsWithFormItemClass.length;

    //add a new sort pair
    var addIcon = utils.container.querySelector('div[class="add-new-sort-pair"] svg');
    expect(addIcon).toMatchSnapshot();
    fireEvent.click(addIcon);

    //check if the new sort pair was added
    elementsWithFormItemClass = utils.container.querySelectorAll(".form-item");
    expect(elementsWithFormItemClass.length).toBe(lengthElementsWithFormItemClass+2);

    //remove added sort pair
    var deleteIcon = utils.container.querySelector('svg[class="MuiSvgIcon-root undefined-sort-pair-clear-icon"]');
    fireEvent.click(deleteIcon);

    //check if deleted
    elementsWithFormItemClass = utils.container.querySelectorAll(".form-item");
    expect(elementsWithFormItemClass.length).toBe(lengthElementsWithFormItemClass);

    const formComponent = utils.container.querySelector('form');

    fireEvent.submit(formComponent);
    expect(onSubmitPropMock).toHaveBeenCalled();
  });

});
