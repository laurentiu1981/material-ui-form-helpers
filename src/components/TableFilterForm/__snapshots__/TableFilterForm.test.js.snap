// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableFilterForm Tests Table Filter Form - add & remove sort pair 1`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_min"
  type="text"
  value="1"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - add & remove sort pair 2`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_max"
  type="text"
  value="10"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - add & remove sort pair 3`] = `
<svg
  aria-hidden="true"
  class="MuiSvgIcon-root"
  focusable="false"
  role="presentation"
  viewBox="0 0 24 24"
>
  <path
    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
  />
</svg>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 1`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="simpleElement"
  type="text"
  value="testInputSimpleElement"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 2`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_min"
  type="text"
  value="1"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 3`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_max"
  type="text"
  value="10"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 4`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="simpleElement"
  type="text"
  value=""
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 5`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_min"
  type="text"
  value=""
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit & cancel 6`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_max"
  type="text"
  value=""
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit 1`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="simpleElement"
  type="text"
  value="testInputSimpleElement"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit 2`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_min"
  type="text"
  value="1"
/>
`;

exports[`TableFilterForm Tests Table Filter Form - submit 3`] = `
<input
  aria-invalid="false"
  class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
  name="intervalElement_max"
  type="text"
  value="10"
/>
`;
